pkg/gpu/stats.go

```diff
@@ -0,0 +1,101 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2024-present Datadog, Inc.
+
+//go:build linux_bpf
+
+package gpu
+
+import (
+	"github.com/DataDog/datadog-agent/pkg/collector/corechecks/gpu/model"
+)
+
+// statsGenerator connects to the active stream handlers and generates stats for the GPU monitoring.
+type statsGenerator struct {
+	streamHandlers      map[model.StreamKey]*StreamHandler // streamHandlers contains the map of active stream handlers.
+	lastGenerationKTime int64                              // lastGenerationTime is the kernel time of the last stats generation.
+	currGenerationKTime int64                              // currGenerationTime is the kernel time of the current stats generation.
+	aggregators         map[uint32]*aggregator             // aggregators contains the map of aggregators
+	sysCtx              *systemContext                     // sysCtx is the system context with global GPU-system data
+}
+
+func newStatsGenerator(sysCtx *systemContext, currKTime int64, streamHandlers map[model.StreamKey]*StreamHandler) *statsGenerator {
+	return &statsGenerator{
+		streamHandlers:      streamHandlers,
+		aggregators:         make(map[uint32]*aggregator),
+		lastGenerationKTime: currKTime,
+		currGenerationKTime: currKTime,
+		sysCtx:              sysCtx,
+	}
+}
+
+func (g *statsGenerator) getStats(nowKtime int64) *model.GPUStats {
+	g.currGenerationKTime = nowKtime
+
+	for key, handler := range g.streamHandlers {
+		aggr := g.getOrCreateAggregator(key)
+		currData := handler.getCurrentData(uint64(nowKtime))
+		pastData := handler.getPastData(true)
+
+		if currData != nil {
+			aggr.processCurrentData(currData)
+		}
+
+		if pastData != nil {
+			aggr.processPastData(pastData)
+		}
+
+		if handler.processEnded {
+			aggr.processEnded = true
+		}
+	}
+
+	g.configureNormalizationFactor()
+
+	stats := model.GPUStats{
+		PIDStats: make(map[uint32]model.PIDStats),
+	}
+
+	for pid, aggregator := range g.aggregators {
+		stats.PIDStats[pid] = aggregator.getStats()
+	}
+
+	g.lastGenerationKTime = g.currGenerationKTime
+
+	return &stats
+}
+
+func (g *statsGenerator) getOrCreateAggregator(streamKey model.StreamKey) *aggregator {
```

Comment:
pass the pid directly, you don't need the whole StreamKey and it is confusing (since the aggregator is per process and not per single stream of a process)

Reply:
but in both cases you don't need the streamID, so pid would be sufficient.
i assume if we want to aggregate by hosts, we would need to introduce a new identifier